playground/backend/cmd/server/controller.go

```diff
@@ -366,29 +351,64 @@ func processCode(ctx context.Context, cacheService cache.Cache, lc *fs_tool.Life
 		}
 	}(successChannel, errorChannel, dataChannel)
 
+	processStep(ctxWithTimeout, pipelineId, cacheService, cancelChan, successChannel, dataChannel, errorChannel, pb.Status_STATUS_RUN_ERROR, pb.Status_STATUS_FINISHED)
+}
+
+// processStep processes each executor's step with cancel and timeout checks.
+// If finishes by canceling, timeout or error - returns false.
+// If finishes successfully returns true.
+func processStep(ctx context.Context, pipelineId uuid.UUID, cacheService cache.Cache, cancelChan, successChan chan bool, dataChan chan interface{}, errorChannel chan error, errorCaseStatus, successCaseStatus pb.Status) bool {
 	select {
-	case <-ctxWithTimeout.Done():
-		finishByContext(ctxWithTimeout, pipelineId, cacheService)
-		return
-	case ok := <-successChannel:
-		data := <-dataChannel
+	case <-ctx.Done():
+		finishByContext(ctx, pipelineId, cacheService)
```

Comment:
Let's rename this function to `finishByTimeout` since that's the only case when this is triggered (you've documented it on the function, which is great - but let's name it as well)

Reply:
Renamed.